### PR TITLE
Extendable diamond cut

### PR DESCRIPTION
**Pull request by amarinkovic**
Builds up on the [upstream](https://github.com/mudgen/diamond-2-hardhat) implementation of `DiamondCutFacet`, extending it with custom code for governance. The idea here is that we only add in custom code before calling the super implementation.

### CODE REVIEW

**Comment by amarinkovic** [comment]
This change is currently constrained by the [gemforge](https://github.com/gemstation/gemforge) DiamondProxy generation code. There is no way to specify a custom `DiamondCutFacet` implementation, in our case that would be the `PhasedDiamondCutFacet`. 

Current version generates a constructor call like this:
```solidity
constructor(address _contractOwner) payable Diamond(_contractOwner, address(new DiamondCutFacet())) {
```

Ideally, it would be possible to generate the `DiamondProxy` such that it passes in a custom facet to the `Diamond` constructor:

```solidity
constructor(address _contractOwner) payable Diamond(_contractOwner, address(new PhasedDiamondCutFacet()))
```

@hiddentao do you think this might be a feature worth including? I guess there could be a config property one could set to override the default implementation.

**Comment by hiddentao** [comment]
Yeah, this is something that could be investigated for Gemforge > https://github.com/gemstation/gemforge/issues/38

File: src/facets/PhasedDiamondCutFacet.sol [hunk 1]
```diff
@@ -6,21 +6,21 @@ pragma solidity 0.8.20;
...
+contract PhasedDiamondCutFacet is DiamondCutFacet {
     /// @notice Add/replace/remove any number of functions and optionally execute
     ///         a function with delegatecall
     /// @param _diamondCut Contains the facet addresses and function selectors
     /// @param _init The address of the contract or facet to execute _calldata
     /// @param _calldata A function call, including function selector and arguments
     ///                  _calldata is executed with delegatecall on _init
-    function diamondCut(FacetCut[] calldata _diamondCut, address _init, bytes calldata _calldata) external override {
+    function diamondCut(FacetCut[] calldata _diamondCut, address _init, bytes calldata _calldata) public override {
         {
```
Comment (kevin-fruitful) [inline src/facets/PhasedDiamondCutFacet.sol]: can we remove this scoping?

**Review by kevin-fruitful — APPROVED** [review]
Left one minor comment. Looks fine otherwise ~

File: src/facets/PhasedDiamondCutFacet.sol (hunk 1 above)
Comment (amarinkovic) [reply src/facets/PhasedDiamondCutFacet.sol]: done

**Comment by hiddentao** [comment]
LGTM

**Pull request merged**